### ISSUE TITLE
Make Cookie Consent Url localizable

### ISSUE THREAD

**Issue by MaiklT** [dnnsoftware/Dnn.Platform#4333]
## Description of problem
On a multi lingual portal it is not possible to enter different Cookie Consent Urls for each language (e.g. [https://en.wikipedia.org/wiki/HTTP_cookie](https://en.wikipedia.org/wiki/HTTP_cookie) for English, [https://de.wikipedia.org/wiki/HTTP-Cookie](https://de.wikipedia.org/wiki/HTTP-Cookie) for German, [https://fr.wikipedia.org/wiki/Cookie_(informatique)](https://fr.wikipedia.org/wiki/Cookie_(informatique)) for French etc.).

## Description of solution
This setting should be dependant from the language.

**Comment by hismightiness** [comment]
Hmmm... Currently, this setting saves to the `PortalSettings` table.  I guess this would mean moving it to `PortalLocalization` table as a new column since there isn't a localized table for portal settings.  

In the specific example of Wikipedia, doesn't it detect and redirect based on browser locale?

**Comment by sleupold** [comment]
@hismightiness 
PortalSettings support localization (since I added to in DNN 7)

**Comment by hismightiness** [comment]
Oh, I missed the `CultureCode` column.  I guess that means this is already a localized setting, right?

**Comment by sleupold** [comment]
yes

**Comment by hismightiness** [comment]
Now I'm confused. 🤔😀

**Comment by sleupold** [comment]
sorry, I meant PortalSettings table supports localization, but for CookieConsent URL, it needs to be wired to the UI.

**Comment by hismightiness** [comment]
But that URL is saved to the `PortalSettings` table.  I guess I'm missing something.  :)

**Comment by sleupold** [comment]
when saving a site setting in PortalSettings table, you may save it localized or "neutral".
In this case, it needs to be change from "neutral" to "localized", allowing diferent URLs per laguage.

**Comment by hismightiness** [comment]
Ah... I didn't check the specific line of code that saves the URL.

**Comment by valadas** [comment]
Well, from a quick look at it every time you change it for any language, it saves the new (same) value for all languages:

![image](https://user-images.githubusercontent.com/6371568/100282543-ce8e3680-2f39-11eb-9a11-20bfaf5606d7.png)

Editing the values in the database appears to work though, so I guess the reading part is fine but we have something wrong in the writing (or maybe modifying part) of the code. it probably modifies all the existing ones instead of just the existing ones that match the currently selected language in the PersonaBar...

**Comment by sleupold** [comment]
@valadas  
this behavior is by design, if update is executed without providing a locale  - with a locale provided, only value for this language will be updated.

**Comment by MaiklT** [comment]
@hismightiness:

> In the specific example of Wikipedia, doesn't it detect and redirect based on browser locale?

No

I tried to add the records in the database - which kinda works, but a bit strange,

To reproduce: Goto [https://www.kaisermaximilianpreis.at](https://www.kaisermaximilianpreis.at).

Click on the link ("Mehr erfahren") - you come to the German Wikipedia Page. Close this page.

Click on "ENG" (top right), then click "Learn more". Again, you come to the German Wikipedia page. Close this page.

Refresh the page. Click "Learn more" again. You come to the English Wikipedia page. Close this page.

Click on "FRA". Click "Apprendre encore plus" - guess what opens. It is the English page again. Close it.

Refresh the page. Click "Apprendre encore plus" again, the French Wikipedia Page opens.

And so on...

???

**Comment by valadas** [comment]
> @valadas
> this behavior is by design, if update is executed without providing a locale - with a locale provided, only value for this language will be updated.

It's not by design, when updating for one language it updates all languages to that value. Something is wrong somewhere. I tdoes read each language value properly but when saving it updates them all to the same value. This is wrong.

**Comment by hismightiness** [comment]
FYI - In general, making an update directly to the database for something like this may not be a true representation of the issue and how to reproduce it so someone can potentially fix it (or even that is does/doesn't work).  I know everyone here so far knows that, so this comment is mostly for anyone else watching.  :)

**Comment by valadas** [comment]
> FYI - In general, making an update directly to the database for something like this may not be a true representation of the issue and how to reproduce it so someone can potentially fix it (or even that is does/doesn't work). I know everyone here so far knows that, so this comment is mostly for anyone else watching. :)

Correct, I was mainly pointing out that after testing this, the logic is fine but the saving is wrong, it updates all languages with the same value no mater which language you are attempting to edit. It is a workaround at least to adjust those existing entries until someone picks up on this issue.

**Comment by sleupold** [comment]
sorry, I misread your statement. the logic should be:
- when updating without providing a CultureCode > update all values (or delete all language specific values and update/create just a neutral one
- when updating providing a CultureCode > update value for this language or create a record, if not exists.

**Comment by valadas** [comment]
Correct

**Comment by sleupold** [comment]
AFAIK still an issue. 
Anyone willing to contribute a PR?

**Comment by hismightiness** [comment]
Still relevant.

**Comment by hismightiness** [comment]
Still relevant.

**Comment by valadas** [comment]
I triaged this, the fix should be pretty similar to https://github.com/dnnsoftware/Dnn.Platform/pull/4851/commits/636f7310355865c5d6715d04562c6d00b15595a3

Basically, we need to read and write using an API that we can provide with the culture code from the request and not let DNN find it from the alias. The alias might not be the language selected for which we want to set the setting.
